File: flet_ble/src/flet_ble/ble.py
"""Python side of the flet-ble extension (see ../../flutter/flet_ble/lib/src/ble.dart)."""

from dataclasses import dataclass, field
from typing import Any, Optional

import flet as ft

__all__ = ["FletBle", "BleScanResultEvent", "BleNotifyEvent", "BleConnectionStateEvent"]


@dataclass
class BleScanResultEvent(ft.Event["FletBle"]):
    """One advertisement seen while scanning."""

    device_id: str
    local_name: str = ""
    rssi: int = 0
    connectable: bool = True
    manufacturer_data: dict[str, str] = field(default_factory=dict)  # company id (decimal str) -> hex
    service_uuids: list[str] = field(default_factory=list)

    @property
    def manufacturer_bytes(self) -> dict[int, bytes]:
        return {int(k): bytes.fromhex(v) for k, v in (self.manufacturer_data or {}).items()}


@dataclass
class BleNotifyEvent(ft.Event["FletBle"]):
    """A notification/indication received from a characteristic."""

    device_id: str
    characteristic: str
    value: str = ""  # hex

    @property
    def bytes(self) -> bytes:
        return bytes.fromhex(self.value or "")


@dataclass
class BleConnectionStateEvent(ft.Event["FletBle"]):
    device_id: str
    connected: bool = False


@ft.control("FletBle")
class FletBle(ft.Service):
    """Minimal BLE central (GATT client) service backed by flutter_blue_plus.

    Add an instance to ``page.services`` and set the ``on_*`` handlers *before*
    adding it (the Flutter side only forwards events that have a handler).
    """

    on_scan_result: Optional[ft.EventHandler[BleScanResultEvent]] = None
    on_notify: Optional[ft.EventHandler[BleNotifyEvent]] = None
    on_connection_state: Optional[ft.EventHandler[BleConnectionStateEvent]] = None
    on_error: Optional[ft.ControlEventHandler["FletBle"]] = None

    # ------------------------------------------------------------ adapter
    async def is_supported(self) -> bool:
        return bool(await self._invoke_method("is_supported"))

    async def adapter_state(self) -> str:
        """'on', 'off', 'unauthorized', 'unavailable', 'turningOn', 'turningOff', 'unknown'."""
        return str(await self._invoke_method("adapter_state"))

    async def turn_on(self) -> bool:
        return bool(await self._invoke_method("turn_on"))

    # --------------------------------------------------------------- scan
    async def start_scan(self, services: Optional[list[str]] = None,
                         manufacturer_ids: Optional[list[int]] = None,
                         timeout_ms: Optional[int] = None, fine_location: bool = True) -> None:
        await self._invoke_method("start_scan", {
            "services": services or [],
            "manufacturer_ids": [int(m) for m in (manufacturer_ids or [])],
            "timeout_ms": timeout_ms,
            "fine_location": fine_location,
        }, timeout=30)

    async def stop_scan(self) -> None:
        await self._invoke_method("stop_scan")

    # ------------------------------------------------------------ connect
    async def connect(self, device_id: str, timeout_ms: int = 20000) -> None:
        await self._invoke_method("connect", {"device_id": device_id, "timeout_ms": timeout_ms},
                                  timeout=timeout_ms / 1000 + 5)

    async def disconnect(self, device_id: str) -> None:
        await self._invoke_method("disconnect", {"device_id": device_id})

    async def is_connected(self, device_id: str) -> bool:
        return bool(await self._invoke_method("is_connected", {"device_id": device_id}))

    async def discover_services(self, device_id: str) -> list[dict[str, Any]]:
        return await self._invoke_method("discover_services", {"device_id": device_id}, timeout=30)

    async def request_mtu(self, device_id: str, mtu: int) -> int:
        return int(await self._invoke_method("request_mtu", {"device_id": device_id, "mtu": mtu}))

    async def mtu(self, device_id: str) -> int:
        return int(await self._invoke_method("mtu", {"device_id": device_id}))

    # --------------------------------------------------------------- gatt
    async def write(self, device_id: str, characteristic: str, data: bytes,
                    without_response: bool = True) -> None:
        await self._invoke_method("write", {
            "device_id": device_id,
            "characteristic": characteristic,
            "data": bytes(data).hex(),
            "without_response": without_response,
        }, timeout=20)

    async def read(self, device_id: str, characteristic: str) -> bytes:
        r = await self._invoke_method("read", {"device_id": device_id, "characteristic": characteristic},
                                      timeout=20)
        return bytes.fromhex(r or "")

    async def set_notify(self, device_id: str, characteristic: str, enable: bool = True) -> bool:
        return bool(await self._invoke_method("set_notify", {
            "device_id": device_id, "characteristic": characteristic, "enable": enable,
        }, timeout=20))

File: flet_ble/src/flutter/flet_ble/lib/src/ble.dart
import 'dart:async';

import 'package:flet/flet.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';

String _hex(List<int> bytes) =>
    bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();

List<int> _unhex(String s) {
  final out = <int>[];
  for (var i = 0; i + 1 < s.length; i += 2) {
    out.add(int.parse(s.substring(i, i + 2), radix: 16));
  }
  return out;
}

/// Flet service wrapping flutter_blue_plus as a plain GATT client.
///
/// Methods (invoked from Python):
///   is_supported, adapter_state, turn_on,
///   start_scan{services, manufacturer_ids, timeout_ms, fine_location},
///   stop_scan, connect{device_id, timeout_ms}, disconnect{device_id},
///   discover_services{device_id}, write{device_id, characteristic, data, without_response},
///   set_notify{device_id, characteristic, enable}, request_mtu{device_id, mtu}
/// Events (to Python):
///   scan_result{device_id, local_name, rssi, connectable, manufacturer_data{id: hex}, service_uuids[]}
///   notify{device_id, characteristic, value(hex)}
///   connection_state{device_id, connected}
///   error(String)
class FletBleService extends FletService {
  FletBleService({required super.control});

  StreamSubscription<List<ScanResult>>? _scanSub;
  StreamSubscription<OnConnectionStateChangedEvent>? _connSub;
  final Map<String, BluetoothDevice> _devices = {};
  final Map<String, List<BluetoothService>> _services = {};
  final Map<String, StreamSubscription<List<int>>> _notifySubs = {};

  @override
  void init() {
    super.init();
    debugPrint("FletBle(${control.id}).init");
    control.addInvokeMethodListener(_invokeMethod);
    _connSub = FlutterBluePlus.events.onConnectionStateChanged.listen((e) {
      final id = e.device.remoteId.str;
      final connected = e.connectionState == BluetoothConnectionState.connected;
      debugPrint("FletBle: $id connected=$connected");
      if (!connected) {
        _dropNotifySubs(id);
        _services.remove(id);
      }
      control.triggerEvent("connection_state", {"device_id": id, "connected": connected});
    });
  }

  void _dropNotifySubs(String deviceId) {
    final keys = _notifySubs.keys.where((k) => k.startsWith("$deviceId/")).toList();
    for (final k in keys) {
      _notifySubs.remove(k)?.cancel();
    }
  }

  BluetoothDevice _device(dynamic args) {
    final id = args["device_id"].toString();
    return _devices.putIfAbsent(id, () => BluetoothDevice.fromId(id));
  }

  Future<BluetoothCharacteristic> _characteristic(dynamic args) async {
    final device = _device(args);
    final id = device.remoteId.str;
    final wanted = Guid(args["characteristic"].toString());
    var services = _services[id];
    if (services == null || services.isEmpty) {
      services = await device.discoverServices();
      _services[id] = services;
    }
    for (final s in services) {
      for (final c in s.characteristics) {
        if (c.uuid == wanted) return c;
      }
    }
    throw Exception("characteristic ${wanted.str128} not found on $id");
  }

  Map<String, dynamic> _scanResultToMap(ScanResult r) {
    final mfg = <String, String>{};
    r.advertisementData.manufacturerData.forEach((k, v) {
      mfg[k.toString()] = _hex(v);
    });
    return {
      "device_id": r.device.remoteId.str,
      "local_name": r.advertisementData.advName.isNotEmpty
          ? r.advertisementData.advName
          : r.device.platformName,
      "rssi": r.rssi,
      "connectable": r.advertisementData.connectable,
      "manufacturer_data": mfg,
      "service_uuids": r.advertisementData.serviceUuids.map((g) => g.str128).toList(),
    };
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("FletBle.$name($args)");
    try {
      switch (name) {
        case "is_supported":
          return await FlutterBluePlus.isSupported;

        case "adapter_state":
          {
            var state = FlutterBluePlus.adapterStateNow;
            if (state == BluetoothAdapterState.unknown) {
              try {
                state = await FlutterBluePlus.adapterState
                    .where((s) => s != BluetoothAdapterState.unknown)
                    .first
                    .timeout(const Duration(seconds: 3));
              } on TimeoutException {
                // keep "unknown"
              }
            }
            return state.name;
          }

        case "turn_on":
          if (!kIsWeb && defaultTargetPlatform == TargetPlatform.android) {
            await FlutterBluePlus.turnOn();
            return true;
          }
          return false;

        case "start_scan":
          {
            await _scanSub?.cancel();
            _scanSub = FlutterBluePlus.onScanResults.listen((results) {
              for (final r in results) {
                control.triggerEvent("scan_result", _scanResultToMap(r));
              }
            }, onError: (Object e) {
              control.triggerEvent("error", "scan: $e");
            });
            final services = ((args["services"] as List?) ?? [])
                .map((s) => Guid(s.toString()))
                .toList();
            final msd = ((args["manufacturer_ids"] as List?) ?? [])
                .map((m) => MsdFilter(int.parse(m.toString())))
                .toList();
            final timeoutMs = args["timeout_ms"];
            await FlutterBluePlus.startScan(
              withServices: services,
              withMsd: msd,
              timeout: timeoutMs != null ? Duration(milliseconds: (timeoutMs as num).toInt()) : null,
              continuousUpdates: true,
              continuousDivisor: 2,
              androidUsesFineLocation: args["fine_location"] == true,
            );
            return true;
          }

        case "stop_scan":
          await FlutterBluePlus.stopScan();
          await _scanSub?.cancel();
          _scanSub = null;
          return true;

        case "connect":
          {
            final device = _device(args);
            final timeoutMs = (args["timeout_ms"] as num?)?.toInt() ?? 20000;
            await device.connect(
              license: License.nonprofit,
              timeout: Duration(milliseconds: timeoutMs),
              mtu: null,
            );
            _services.remove(device.remoteId.str);
            return true;
          }

        case "disconnect":
          {
            final device = _device(args);
            _dropNotifySubs(device.remoteId.str);
            _services.remove(device.remoteId.str);
            await device.disconnect();
            return true;
          }

        case "is_connected":
          return _device(args).isConnected;

        case "discover_services":
          {
            final device = _device(args);
            final services = await device.discoverServices();
            _services[device.remoteId.str] = services;
            return services
                .map((s) => {
                      "uuid": s.uuid.str128,
                      "characteristics": s.characteristics
                          .map((c) => {
                                "uuid": c.uuid.str128,
                                "read": c.properties.read,
                                "write": c.properties.write,
                                "write_without_response": c.properties.writeWithoutResponse,
                                "notify": c.properties.notify,
                                "indicate": c.properties.indicate,
                              })
                          .toList(),
                    })
                .toList();
          }

        case "request_mtu":
          {
            final device = _device(args);
            return await device.requestMtu((args["mtu"] as num).toInt());
          }

        case "mtu":
          return _device(args).mtuNow;

        case "write":
          {
            final c = await _characteristic(args);
            await c.write(_unhex(args["data"].toString()),
                withoutResponse: args["without_response"] != false);
            return true;
          }

        case "read":
          {
            final c = await _characteristic(args);
            return _hex(await c.read());
          }

        case "set_notify":
          {
            final c = await _characteristic(args);
            final id = c.remoteId.str;
            final key = "$id/${c.uuid.str128}";
            final enable = args["enable"] != false;
            await _notifySubs.remove(key)?.cancel();
            if (enable) {
              _notifySubs[key] = c.onValueReceived.listen((value) {
                control.triggerEvent("notify", {
                  "device_id": id,
                  "characteristic": c.uuid.str128,
                  "value": _hex(value),
                });
              });
            }
            return await c.setNotifyValue(enable);
          }

        default:
          throw Exception("Unknown FletBle method: $name");
      }
    } catch (e) {
      debugPrint("FletBle.$name failed: $e");
      throw _FletBleException("$name: $e");
    }
  }

  @override
  void dispose() {
    debugPrint("FletBle(${control.id}).dispose()");
    control.removeInvokeMethodListener(_invokeMethod);
    _scanSub?.cancel();
    _connSub?.cancel();
    for (final s in _notifySubs.values) {
      s.cancel();
    }
    _notifySubs.clear();
    super.dispose();
  }
}

class _FletBleException implements Exception {
  final String message;
  const _FletBleException(this.message);

  @override
  String toString() => message;
}
